### PR TITLE
Recognize voice messages

### PR DESCRIPTION
**Pull request by AlexanderSutul**
A boilerplate of voice recognition was added.
There is todo: add converting from .oga to .wav to make recognition using google speech correctly.

### CODE REVIEW

File: bot/utils/audio_recognition.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+from google.cloud import speech
+import requests
+import logging
+import json
+import io
+from pydub import AudioSegment
+
+TOKEN = os.environ["TOKEN"]
+LANG = os.environ["LANG"]
+SAMPLE_RATE_HERTZ = os.environ["SAMPLE_RATE_HERTZ"]
```
Comment (C-Pro) [inline bot/utils/audio_recognition.py:11]: It is ok to fix these settings as a constants right here. No need to put them in env.

Comment (AlexanderSutul) [reply bot/utils/audio_recognition.py:11]: Hi! No problem, I'll fix it this evening. How about converting from .oga to .wav? Any ideas?

Comment (C-Pro) [reply bot/utils/audio_recognition.py:11]: ffmpeg filexxx.oga /tmp/blabla/filexxx.wav

Comment (C-Pro) [reply bot/utils/audio_recognition.py:11]: We will need to add ffmpeg to docker image and mount some temp directory from host.

**Comment by AlexanderSutul** [comment]
@C-Pro hi! Sorry for this long feature implementation, I was very busy at work. So the feature is ready.

File: bot/utils/voice_recognition.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+speechClient = speech.SpeechClient()
+
+
+def _get_audio_file_data(file_id):
+    """
+    gets audio file 
+    """
+    url = f"https://api.telegram.org/bot{TOKEN}/getFile?file_id={file_id}"
+    response = requests.get(url)
+    jsonResponse = response.json()
```
Comment (egregors) [inline bot/utils/voice_recognition.py:28]: Variable in function should be [lowercase](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)

File: bot/utils/voice_recognition.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+LANG = "ru-RU"
+SAMPLE_RATE_HERTZ = 48000
+ENCODING_TYPE = "LINEAR16"
+CHANNEL_COUNT = 1
+ENABLE_AUTOMATIC_PUNCTUATION = True
+
+logger = logging.getLogger(__name__)
+
+speechClient = speech.SpeechClient()
```
Comment (egregors) [inline bot/utils/voice_recognition.py:19]: Variable in function should be [lowercase](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)

File: bot/utils/voice_recognition.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def _get_audio_file_data(file_id):
+    """
+    gets audio file 
+    """
+    url = f"https://api.telegram.org/bot{TOKEN}/getFile?file_id={file_id}"
+    response = requests.get(url)
+    jsonResponse = response.json()
+    logger.info(json.dumps(jsonResponse, indent=4))
```
Comment (egregors) [inline bot/utils/voice_recognition.py:29]: Logs shouldn't be printed without context

File: bot/utils/voice_recognition.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    return converted_wav_bin
+
+
+def _send_binary_to_google_speech(content):
+    """
+    sends binary data of voice to google speech
+    """
+    config = {
+        "language_code": LANG,
+        "sample_rate_hertz": int(SAMPLE_RATE_HERTZ),
```
Comment (egregors) [inline bot/utils/voice_recognition.py:67]: Redundant typecasting

File: bot/utils/voice_recognition.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        "model": "default",
+    }
+    audio = {"content": content}
+    speech_response = speechClient.recognize(config=config, audio=audio)
+    return speech_response
+
+
+def _check_google_speech_response(response):
+    result = response.results[-1]
+    logger.info(result)
```
Comment (egregors) [inline bot/utils/voice_recognition.py:80]: Logging without context

File: bot/utils/voice_recognition.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    gets text from voice
+    """
+    try:
+        file = _get_audio_file_data(file_id)
+        content = _get_binary_content(file)
+        response = _send_binary_to_google_speech(content)
+        result = _check_google_speech_response(response)
+        logger.info(f"{result}")
+        return result
+    except Exception as ex:
```
Comment (egregors) [inline bot/utils/voice_recognition.py:99]: Too common Exception type. Is it possible to define more particular ones?

File: bot/utils/voice_recognition.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+def _get_binary_content(file):
+    file_name = file["file_path"]
+    url = f"https://api.telegram.org/file/bot{TOKEN}/{file_name}"
+    response = requests.get(url, stream=True)
+
+    suffix = ".oga"
+    tmp_voice = NamedTemporaryFile(delete=False, suffix=suffix)
+    tmp_voice.write(response.content)
+    tmp_voice_filename_ogg = tmp_voice.name
+    logger.info(f"Name of tempfile {tmp_voice_filename_ogg}")
```
Comment (egregors) [inline bot/utils/voice_recognition.py:43]: Don't use `f-strings` in logging: 

> Formatting of these arguments is deferred until it cannot be avoided. That means the final message is not evaluated if its log level is below the logger’s log level. On the other hand, f-string is really an expression that evaluated at runtime and it lacks logging’s optimizations.

**Review by egregors — CHANGES_REQUESTED** [review]
Looks pretty cool! 🔥 🔥 🔥 🔥 🔥  Some cleanup here and we'll merge it in the dev branch I guess. It should be tested a bit in devs chat. @AlexanderSutul r u in dev chat already?

@C-Pro could you please add new ENVs vars in dev stage environment?

File: bot/utils/voice_recognition.py (hunk 2 above)
Comment (AlexanderSutul) [reply bot/utils/voice_recognition.py:28]: Fixed.

File: bot/utils/voice_recognition.py (hunk 3 above)
Comment (AlexanderSutul) [reply bot/utils/voice_recognition.py:19]: Fixed.

File: bot/utils/voice_recognition.py (hunk 4 above)
Comment (AlexanderSutul) [reply bot/utils/voice_recognition.py:29]: Fixed.

File: bot/utils/voice_recognition.py (hunk 5 above)
Comment (AlexanderSutul) [reply bot/utils/voice_recognition.py:67]: Fixed.

File: bot/utils/voice_recognition.py (hunk 6 above)
Comment (AlexanderSutul) [reply bot/utils/voice_recognition.py:80]: Fixed

File: bot/utils/voice_recognition.py (hunk 8 above)
Comment (AlexanderSutul) [reply bot/utils/voice_recognition.py:43]: Removed. It was a debug info.

File: bot/utils/voice_recognition.py (hunk 7 above)
Comment (AlexanderSutul) [reply bot/utils/voice_recognition.py:99]: I'm not a full python developer😄  Probably you could give me an advice, which approach to use here to resolve this discussion?

**Review by egregors — APPROVED** [review]
LGFM

File: bot/utils/voice_recognition.py (hunk 7 above)
Comment (egregors) [reply bot/utils/voice_recognition.py]: Usually, you should define which exceptions are expectable, and try to catch them. An unexpectable exception must be raised to be caught outside a current abstraction layer.

Comment (AlexanderSutul) [reply bot/utils/voice_recognition.py]: Done, please check.

Comment (egregors) [reply bot/utils/voice_recognition.py]: 🔥 🔥 🔥 🔥 🔥 🔥

**Review by egregors — APPROVED** [review]

**Comment by egregors** [comment]
@AlexanderSutul thanks a lot for your contributing! 👍 

Let's test it a bit in a dev chat, then I gonna merge it into ~~master~~ main branch

**Comment by AlexanderSutul** [comment]
> @AlexanderSutul thanks a lot for your contributing! 👍
> 
> Let's test it a bit in a dev chat, then I gonna merge it into ~master~ main branch

Could you add me to the dev chat?

**Comment by egregors** [comment]
@AlexanderSutul sure! Text me your tg nickname please, or just answer me in vldc chat

**Pull request merged**